common/setups/returnn_common/nnet_constructor.py

```diff
@@ -0,0 +1,425 @@
+    """
+    A helper class to store input args for a nn.Dim object
+
+    :param name: name of the dim
+    :param dim: dimension size (feature axis size or label index count for sparse_dim), None for spatial axis
+    """
+
+    name: str
+    dim: Optional[Union[int, tk.Variable]]
+    is_feature: bool = False
+
+
+@dataclass(frozen=True)
+class DataInitArgs:
+    """
+    A helper class to store needed input args for a nn.Data object without `returnn_common` dependency
+
+    :param name: name of the data (equivalent to the extern_data entry)
+    :param available_for_inference: if this data is available during decoding/forward pass etc...
+    :param dim_tags: list of dim tags representing an axis of the data, without batch or a hidden sparse dim
+    :param sparse_dim: provide this dim to make the data sparse and define
+    :param has_batch_dim: set to False do create data without a batch dim (probably never needed)
+    """
+
+    name: str
+    available_for_inference: bool
+    dim_tags: List[DimInitArgs]
+    sparse_dim: Optional[DimInitArgs]
+    has_batch_dim: bool = True
+
+    def __post_init__(self):
+        if self.sparse_dim is not None:
+            assert (
+                self.sparse_dim.dim is not None
+            ), "A sparse dim can not have 'None' as dimension"
+            assert (
+                self.sparse_dim.is_feature is True
+            ), "A sparse dim should be a feature dim"
+
+    def _sis_hash(self) -> bytes:
+        # INFO: asdict is recursive, so DimInitArgs will be converted as well
+        return sis_hash_helper(asdict(self))
+
+
+class ReturnnCommonSerializer(DelayedBase):
+    """
+    A helper class to serialize a Returnn config with returnn_common elements.
+    Should be passed to either `returnn_prolog` or `returnn_epilog` in a `ReturnnConfig`
+
+    The tasks are:
+     - managing the returnn_common version (via sys.path)
+     - managing returnn_common net definitions (via importing a nn.Module class definition, using `ReturnnCommonImport`)
+     - managing returnn_common net construction which returns the final net, using `ReturnnCommonImport`
+       (via importing a (epoch, **kwargs) -> nn.Module function)
+     - managing nn.Dim/Data and the extern_data entry via `ReturnnCommonExternData`, `DimInitArgs` and `DataInitArgs`
+     - managing the package imports from which all imports can be found
+     - optionally make a local copy of all imported code instead if importing it directly out of the recipes
+
+    """
+
+    def __init__(
+        self,
+        serializer_objects: List[SerializerObject],
+        returnn_common_root: Optional[tk.Path] = None,
+        packages: Optional[List[Union[str, tk.Path]]] = None,
+        make_local_package_copy=False,
+    ):
+        """
+
+        :param serializer_objects: all serializer objects which are serialized into a string  in order
+        :param returnn_common_root:
+        :param packages:
+        :param make_local_package_copy:
```

Comment:
I think it would be good if this is documented.
```suggestion
        :param returnn_common_root: Path to returnn_common, if None, assumes direct import is fine
        :param packages: Path to packages to import, if None, tries to extract them from serializer_objects
        :param make_local_package_copy: whether to make a local copy of imported code into the Job directory 
```